docs/proposals/connector-protocol-v2-addendum-migration.md

```diff
@@ -0,0 +1,127 @@
+  |      +----------> | Error  | <-----------+---| Migrating |
+  |      |            +--------+   |         |   +-----------+
+  |      |                |        |         |         |
+  |      |                V        V         |         |
+  |      |       +-------------------+       |         |
+  |      +------>+  Disconnected     |<------+---------+
+  |              +-------------------+                 |
+  |                                                    |
+  +----------------------------------------------------+
+```
+
+In the connected state the following frame types are valid:
+
+    - HELLO: Connector -> Worker (next state = handshake)
+    - ERROR: * -> * (next state = error)
+    - MIGRATE: Worker -> Connector
+
+In the handshake state the following frame types are valid:
+
+    - OK: Worker -> Connector (next state = streaming)
+    - ERROR: * -> * (next state = error)
+    - MIGRATE: Worker -> Connector (next state = migrating)
+
+In the streaming state the following frame types are valid:
+
+   - NOFITY: Connector -> Worker
+   - NOTIFY_ACK: Worker -> Connector
+   - MESSAGE: Connector -> Worker
+   - ACK: Worker -> Connector
+   - RESTART: Worker -> Connector
+   - ERROR: * -> * (next state = error)
+   - MIGRATE: Worker -> Connector (next state = migrating)
+
+In the migrating state, the following frame types are valid:
+  - MIGRATE_OK: Connector -> Worker (next state = connected)
+  - ERROR: * -> * (next state = error)
+
+
+## Migration Protocol
+
+In order to accomplish a migration, the following sequence of event needs to occur:
+
+| order | Connector | Worker |  Notes |
+| - | - | - | - |
+| 0 | Connect | - | Connector is in one of {Connected, Handshake, Steaming} |
+| 1 | - | send MIGRATE | This message contains information about the new target address |
+| 2 | close all streams | - | - |
+| 3 | send EOS for all streams | - | - |
+| 4 | - | checkpoint all streams | - |
+| 5 | - | migrate source state to new worker | - |
+| 6 | - | Ack all streams | - |
+| 7 | connect to new target address | - | Old connection is still open |
+| 8 | Handshake on new address | - | Old connection is still open |
+| 9 | send MIGRATION_ACK | - | Send this on old connection |
+| 10 | - | Worker closes connection, completes shut down | - |
+| 11 | resume streaming | - | on new connection; Send NOTIFYs, etc. | - |
+
+### Notes
+- The Connector needs to connect to the new source before closing the old one
+- In the case of ERROR during this process, the connector will have enough information to reach out to BOTH the old and the new source addresses... IF attempting to reconnect after an error or loss of connection, the Connector may connect to either, and if no reponse, try the other one. If both addresse are live and the Connector connected to the wrong one, it can send a MIGRATE message to instruct the Connector to connect to the correct address.
+- The old Worker remains alive until the the migration is complete.
```

Comment:
The old worker does stay alive until migration is complete.  During a shrink, the leaving workers only shut down once all state has been migrated.  This is already the current behavior.  See my question in the `Conversation` section about what to do if external connector dies (times out from Wallaroo's perspective).